osbuild.py

```diff
@@ -196,6 +193,60 @@ def print_header(title, options, machine_name):
     print(f"\t# nsenter -a --wd=/root -t `machinectl show {machine_name} -p Leader --value`")
     print()
 
+class Stage:
+    def __init__(self, stage, objects):
+        m = hashlib.sha256()
+        m.update(json.dumps(stage, sort_keys=True).encode())
+
+        self.name = stage["name"]
+        self.options = stage.get("options", {})
+        self.id = m.hexdigest()
+        self.base = stage.get("base")
+        self.resources = stage.get("systemResourcesFromEtc", [])
+        self.objects = objects
+
+    def run(self, interactive=False):
+        output_tree = os.path.join(self.objects, self.id)
+
+        if os.path.exists(output_tree):
+            return
+
+        with BuildRoot() as buildroot, tmpfs() as tree:
+            if interactive:
+                print_header(f"{self.name} ({self.id})", self.options, buildroot.machine_name)
+            if self.base:
+                input_tree = os.path.join(self.objects, self.base)
+                subprocess.run(["cp", "-a", f"{input_tree}/.", tree], check=True)
+            r = buildroot.run_stage(self.name, self.options, self.resources, tree, interactive)
+            shutil.rmtree(output_tree, ignore_errors=True)
+            os.makedirs(output_tree, mode=0o755)
+            subprocess.run(["cp", "-a", f"{tree}/.", output_tree], check=True)
+
+        return r
+
```

Comment:
Nitpick: Please leave two blank lines between top-level statements for consistency. (Sorry!)

Reply:
Noted. And apology accepted.